### PR TITLE
ci: split tests into five deterministic batches

### PR DESCRIPTION
**Pull request by Lykhoyda**
## Intent

Implement ask-llm GitHub issue 252: split the repository test suite into exactly five deterministic parallel CI batches using a visible native GitHub Actions matrix. Assign every current test to exactly one batch in deterministic order with no omissions or duplicate execution from glob overlap, and remain correct as tests are added without a hand-maintained file list. Preserve the existing supported Node/platform setup, immutable dependency installation behavior, environment, timeouts, coverage or required artifacts, failure semantics, and branch-protection check contract. Preserve per-Node/OS required-check diagnostic semantics so each of the three legacy check names reports only its matching five batches rather than coupling all names to the full 15-leg matrix. Run dependency install, build, lint, and the shared-changeset guard only once per Node/OS setup leg, then fan out only test execution through bounded native Actions artifact composition rather than a custom scheduler. Aggregate matching result artifacts only after all five batches complete. Use existing repository tooling rather than a general scheduler or bespoke CI framework. Add focused sharding and workflow-contract tests covering complete coverage, no overlap, stable assignment, empty and small-suite behavior, command quoting, one-time setup, exact five-way fan-out, and per-platform check isolation. Validate workflow syntax and the same commands every matrix batch runs. Keep concise contributor and architectural documentation current. Update the existing unmerged PR 259, link issue 252, explain the invariant and evidence no tests were lost, and address both review threads with exact evidence. Do not merge.

## What Changed

- Split each supported Node/platform CI leg into one shared setup job and five native matrix test batches, with isolated result aggregation under the existing required-check names.
- Add Vitest-based test discovery and deterministic round-robin batching, including safe handling for empty batches and cross-platform command invocation.
- Add focused sharding and workflow-contract tests, plus contributor guidance and an ADR documenting the five-batch invariant.

## Risk Assessment

✅ Low: The prior full-matrix dependency coupling is removed, and the three independent setup, five-batch, and aggregation chains now satisfy the source-verifiable sharding and legacy-check contracts.

## Testing

Focused contracts and YAML syntax passed; after recovering the documented build prerequisite, a real CI batch passed, and live discovery proved all 70 legacy tests plus two focused tests form five stable batches of 15/15/14/14/14 with no omissions or duplicates.

<details>
<summary>Evidence: Successful CI batch transcript</summary>

```text
Test batch 1/5: 15/72 files
  packages/antigravity-mcp/src/__tests__/integration.test.ts
  packages/antigravity-mcp/src/tools/__tests__/toolSchemas.test.ts
  packages/claude-mcp/src/__tests__/integration.test.ts
  packages/claude-plugin/src/__tests__/auto-pause.test.ts
  packages/claude-plugin/src/__tests__/debounce-state.test.ts
  packages/claude-plugin/src/__tests__/stop-gate.test.ts
  packages/codex-mcp/src/tools/__tests__/toolSchemas.test.ts
  packages/gemini-mcp/src/__tests__/tier-executor.test.ts
  packages/gemini-mcp/src/utils/__tests__/geminiStreamJsonl.test.ts
  packages/llm-mcp/src/__tests__/multiLlm.test.ts
  packages/ollama-mcp/src/tools/__tests__/toolContract.test.ts
  packages/shared/src/__tests__/commandExecutor.test.ts
  packages/shared/src/__tests__/providers.test.ts
  packages/shared/src/__tests__/shellPath.test.ts
  scripts/run-test-batch.test.mjs

 RUN  v4.1.0 /Users/antonlykhoyda/.no-mistakes/worktrees/ae26ae07ac5b/01KZ6G4ERQMB2X08KDPSK93VDM


 Test Files  13 passed | 2 skipped (15)
      Tests  197 passed | 2 skipped (199)
   Start at  16:18:54
   Duration  2.91s (transform 4.22s, setup 0ms, import 5.83s, tests 6.39s, environment 2ms)
```
</details>
<details>
<summary>Evidence: Five-batch coverage proof</summary>

```text
ask-llm deterministic CI partition proof
Current root discovery: 72 files
Legacy package discovery: 70 files
Focused CI script tests added: 2 files
Legacy files omitted: 0
Unexpected package files: 0
Five-batch union size: 72
Duplicate assignments: 0
Omitted assignments: 0
Stable under reversed discovery order: true
Batch sizes: 15, 15, 14, 14, 14

Batch 1/5
  packages/antigravity-mcp/src/__tests__/integration.test.ts
  packages/antigravity-mcp/src/tools/__tests__/toolSchemas.test.ts
  packages/claude-mcp/src/__tests__/integration.test.ts
  packages/claude-plugin/src/__tests__/auto-pause.test.ts
  packages/claude-plugin/src/__tests__/debounce-state.test.ts
  packages/claude-plugin/src/__tests__/stop-gate.test.ts
  packages/codex-mcp/src/tools/__tests__/toolSchemas.test.ts
  packages/gemini-mcp/src/__tests__/tier-executor.test.ts
  packages/gemini-mcp/src/utils/__tests__/geminiStreamJsonl.test.ts
  packages/llm-mcp/src/__tests__/multiLlm.test.ts
  packages/ollama-mcp/src/tools/__tests__/toolContract.test.ts
  packages/shared/src/__tests__/commandExecutor.test.ts
  packages/shared/src/__tests__/providers.test.ts
  packages/shared/src/__tests__/shellPath.test.ts
  scripts/run-test-batch.test.mjs

Batch 2/5
  packages/antigravity-mcp/src/__tests__/machine-options.test.ts
  packages/antigravity-mcp/src/utils/__tests__/agyVersion.test.ts
  packages/claude-mcp/src/__tests__/nested-guard.test.ts
  packages/claude-plugin/src/__tests__/codex-pair-debounce-worker.test.ts
  packages/claude-plugin/src/__tests__/manifest.test.ts
  packages/codex-mcp/src/__tests__/integration.test.ts
  packages/codex-mcp/src/utils/__tests__/codexDoctor.test.ts
  packages/gemini-mcp/src/__tests__/tier-guidance.test.ts
  packages/llm-mcp/src/__tests__/availability.test.ts
  packages/llm-mcp/src/__tests__/repl.test.ts
  packages/ollama-mcp/src/utils/__tests__/ollamaExecutor.test.ts
  packages/shared/src/__tests__/commandExecutor.timeout.test.ts
  packages/shared/src/__tests__/registry.test.ts
  packages/shared/src/__tests__/usage.test.ts
  scripts/test-workflow-contract.test.mjs

Batch 3/5
  packages/antigravity-mcp/src/__tests__/smoke.test.ts
  packages/antigravity-mcp/src/utils/__tests__/antigravityExecutor.test.ts
  packages/claude-mcp/src/__tests__/smoke.test.ts
  packages/claude-plugin/src/__tests__/codex-pair-prompt-drain.test.ts
  packages/claude-plugin/src/__tests__/providers.test.ts
  packages/codex-mcp/src/__tests__/machine-options.test.ts
  packages/codex-mcp/src/utils/__tests__/codexExecutor.test.ts
  packages/gemini-mcp/src/tools/__tests__/toolContract.test.ts
  packages/llm-mcp/src/__tests__/constants.test.ts
  packages/llm-mcp/src/__tests__/smoke.test.ts
  packages/shared/src/__tests__/askResponse.antigravity.test.ts
  packages/shared/src/__tests__/doctor.test.ts
  packages/shared/src/__tests__/responseCache.test.ts
  packages/shared/src/changeMode/__tests__/changeModeChunker.test.ts

Batch 4/5
  packages/antigravity-mcp/src/tools/__tests__/simple-tools.test.ts
  packages/antigravity-mcp/src/utils/__tests__/invocationLock.test.ts
  packages/claude-mcp/src/tools/__tests__/toolContract.test.ts
  packages/claude-plugin/src/__tests__/codex-pair-session.test.ts
  packages/claude-plugin/src/__tests__/session-registry.test.ts
  packages/codex-mcp/src/__tests__/quota-detection.test.ts
  packages/gemini-mcp/src/__tests__/integration.test.ts
  packages/gemini-mcp/src/utils/__tests__/convertFileSyntax.test.ts
  packages/llm-mcp/src/__tests__/machine-cli.test.ts
  packages/ollama-mcp/src/__tests__/integration.test.ts
  packages/shared/src/__tests__/askResponse.test.ts
  packages/shared/src/__tests__/machine.test.ts
  packages/shared/src/__tests__/serverFactory.test.ts
  packages/shared/src/changeMode/__tests__/changeModeParser.test.ts

Batch 5/5
  packages/antigravity-mcp/src/tools/__tests__/toolContract.test.ts
  packages/antigravity-mcp/src/utils/__tests__/transcriptReader.test.ts
  packages/claude-mcp/src/utils/__tests__/claudeExecutor.test.ts
  packages/claude-plugin/src/__tests__/codex-pair-watch.test.ts
  packages/claude-plugin/src/__tests__/skills-and-agents.test.ts
  packages/codex-mcp/src/__tests__/smoke.test.ts
  packages/gemini-mcp/src/__tests__/smoke.test.ts
  packages/gemini-mcp/src/utils/__tests__/geminiExecutor.test.ts
  packages/llm-mcp/src/__tests__/machine.test.ts
  packages/ollama-mcp/src/__tests__/smoke.test.ts
  packages/shared/src/__tests__/chunkCache.test.ts
  packages/shared/src/__tests__/progressTracker.test.ts
  packages/shared/src/__tests__/sessions.test.ts
  packages/shared/src/changeMode/__tests__/changeModeTranslator.test.ts
```
</details>

## Pipeline

Updates from [git push no-mistakes](https://github.com/kunchenguid/no-mistakes)

<details>
<summary>✅ **intent** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Rebase** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>🔧 **Review** - 1 issue found → auto-fixed ✅</summary>

- 🚨 `.github/workflows/ci.yml:98` - Required criterion: each legacy check must report only its matching five batches “rather than coupling all names to the full 15-leg matrix.” Every gate leg has `needs: test-batches`, so GitHub waits for the entire 15-leg matrix: if Ubuntu’s five batches finish while a Windows batch hangs, both Ubuntu checks remain pending. The contract test misses this because it checks artifact patterns but not dependency isolation. Split setup, five-batch fan-out, and aggregation into independent Node/OS chains, or explicitly relax the requirement.

🔧 Fix: Isolate CI batch gates by Node and platform
✅ Re-checked - no issues remain.

</details>

<details>
<summary>✅ **Test** - passed</summary>

✅ No issues found.
- <code>Inspected `7cf7d6200ee9cd00ec093f7fa8b13d32278eb578..5bd342677b3c8934b8f4d42a9598b31b495ba99e` and live Vitest discovery.</code>
- `yarn vitest run scripts/run-test-batch.test.mjs scripts/test-workflow-contract.test.mjs`
- `ruby -e "require 'yaml'; YAML.load_file('.github/workflows/ci.yml')"`
- <code>`yarn build` to recover the required workspace setup.</code>
- <code>`yarn test:batch 1/5` using the exact CI batch entry point.</code>
- `Explicit legacy-workspace versus root discovery comparison and five-manifest composition.`

</details>

<details>
<summary>✅ **Document** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Lint** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Push** - passed</summary>

✅ No issues found.

</details>

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -56,13 +59,33 @@ jobs:
     - name: Lint
       run: yarn lint
 
-    - name: Run tests
-      run: yarn test
+    - name: Run test batch ${{ matrix.batch }}/5
+      run: yarn test:batch "${{ matrix.batch }}/5"
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply .github/workflows/ci.yml:148]: Fixed in 3c07f8a. Each Node/OS chain now runs `yarn install --immutable`, `yarn build`, `yarn lint`, and the ADR-119 changeset guard exactly once in its `test-setup-*` job. That job uploads a one-day platform artifact; only the five `test-batches-*` legs fan out, restoring that artifact and running `yarn test:batch N/5`, then publishing bounded success markers for their matching gate. `scripts/test-workflow-contract.test.mjs` proves setup contains all four commands and every batch job contains none of them. Live evidence: all three setup jobs and all fifteen test-only batch jobs passed in the 25/25 green PR run.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -56,13 +59,33 @@ jobs:
...
+        node-version: [22.x, 24.x]
+        os: [ubuntu-latest]
+        include:
+          - node-version: 22.x
+            os: windows-latest
+    steps:
+    - name: Require every test batch
+      env:
+        TEST_BATCH_RESULT: ${{ needs.test-batches.result }}
+      run: test "$TEST_BATCH_RESULT" = success
```
[the remark this answers is not in the document]
Comment (Lykhoyda) [reply .github/workflows/ci.yml]: Fixed in 3c07f8a. The workflow now has three independent native chains: `test-setup-node22-ubuntu` → `test-batches-node22-ubuntu` → `test-node22-ubuntu`, the equivalent Node 24 Ubuntu chain, and the Node 22 Windows chain. Each legacy check has `needs:` only on its own five-leg batch matrix, so another platform can fail or hang without changing this check’s result or readiness. `scripts/test-workflow-contract.test.mjs` asserts each gate’s exact matching dependency and rejects references to either other chain. Live evidence: all three historical checks (`test (22.x, ubuntu-latest)`, `test (24.x, ubuntu-latest)`, and `test (22.x, windows-latest)`) passed independently in the 25/25 green PR run.
